### ISSUE TITLE
Segmentation fault in use and in tests

### ISSUE THREAD

**Issue by CryptArchy** [sfackler/rust-openssl#413]
After struggling with the usual Windows build issues, I finally got my own project (which depends on this library) to build using the trick of setting:

```
export OPENSSL_LIB_DIR=C:/OpenSSL-Win64/lib
export OPENSSL_INCLUDE_DIR=C:/OpenSSL-Win64/include
export OPENSSL_LIBS=ssleay32:libeay32
```

When I tried to run it though, I imediately got a `Segmentation Fault`. 
So I cloned this repo to try and nail it down.  I was able to build this library directly, but got the segmentation fault when running tests. I've copied the output below, as I'm not sure where to go from here. It seems like maybe libcrypto isn't working right?

I installed the windows version of OpenSSL linked to from the readme.

```
$ cargo test --no-fail-fast
     Running target\debug\openssl-12ac8d499349ad2b.exe

running 105 tests
   Doc-tests openssl

running 22 tests
test bn::BigNum::checked_shl1_1 ... FAILED
test bn::BigNum::checked_shl1_0 ... FAILED
test bn::BigNum::abs_cmp_0 ... FAILED
test bn::BigNum::checked_mod_add_0 ... FAILED
test bn::BigNum::new_from_slice_0 ... FAILED
test bn::BigNum::negate_0 ... FAILED
test bn::BigNum::checked_sqr_0 ... FAILED
test bn::BigNum::to_dec_str_0 ... FAILED
test bn::BigNum::to_vec_0 ... FAILED
test bn::BigNum::to_hex_str_0 ... FAILED
test crypto::hash::Hasher_0 ... FAILED
test crypto::hash::Hasher_1 ... FAILED
test crypto::hmac::HMAC_0 ... FAILED
test x509::X509Generator::add_extension_0 ... ok
test crypto::hmac::HMAC_1 ... FAILED
test x509::X509Generator::add_extensions_0 ... ok
test x509::X509Generator::add_name_0 ... ok
test x509::X509Generator::add_names_0 ... ok
test x509::extension::AltNameOption::Other_0 ... ok
test x509::X509Generator_0 ... FAILED
test x509::extension::Extension::OtherNid_0 ... ok
test x509::extension::Extension::OtherStr_0 ... ok

failures:

---- bn::BigNum::checked_shl1_1 stdout ----
        thread 'bn::BigNum::checked_shl1_1' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::checked_shl1_0 stdout ----
        thread 'bn::BigNum::checked_shl1_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::abs_cmp_0 stdout ----
        thread 'bn::BigNum::abs_cmp_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- bn::BigNum::checked_mod_add_0 stdout ----
        thread 'bn::BigNum::checked_mod_add_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::new_from_slice_0 stdout ----
        thread 'bn::BigNum::new_from_slice_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::negate_0 stdout ----
        thread 'bn::BigNum::negate_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::checked_sqr_0 stdout ----
        thread 'bn::BigNum::checked_sqr_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::to_dec_str_0 stdout ----
        thread 'bn::BigNum::to_dec_str_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::to_vec_0 stdout ----
        thread 'bn::BigNum::to_vec_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- bn::BigNum::to_hex_str_0 stdout ----
        thread 'bn::BigNum::to_hex_str_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- crypto::hash::Hasher_0 stdout ----
        thread 'crypto::hash::Hasher_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- crypto::hash::Hasher_1 stdout ----
        <anon>:8:1: 8:22 warning: unused result which must be used, #[warn(unused_must_use)] on by default
<anon>:8 h.write_all(data[0]);
         ^~~~~~~~~~~~~~~~~~~~~
<anon>:9:1: 9:22 warning: unused result which must be used, #[warn(unused_must_use)] on by default
<anon>:9 h.write_all(data[1]);
         ^~~~~~~~~~~~~~~~~~~~~
thread 'crypto::hash::Hasher_1' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- crypto::hmac::HMAC_0 stdout ----
        thread 'crypto::hmac::HMAC_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- crypto::hmac::HMAC_1 stdout ----
        <anon>:10:1: 10:22 warning: unused result which must be used, #[warn(unused_must_use)] on by default
<anon>:10 h.write_all(data[0]);
          ^~~~~~~~~~~~~~~~~~~~~
<anon>:11:1: 11:22 warning: unused result which must be used, #[warn(unused_must_use)] on by default
<anon>:11 h.write_all(data[1]);
          ^~~~~~~~~~~~~~~~~~~~~
thread 'crypto::hmac::HMAC_1' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312

---- x509::X509Generator_0 stdout ----
        <anon>:6:5: 6:25 warning: unused import, #[warn(unused_imports)] on by default
<anon>:6 use std::io::prelude::*;
             ^~~~~~~~~~~~~~~~~~~~
<anon>:7:5: 7:20 warning: unused import, #[warn(unused_imports)] on by default
<anon>:7 use std::path::Path;
             ^~~~~~~~~~~~~~~
<anon>:2:9: 2:17 warning: unknown lint: `unstable`, #[warn(unknown_lints)] on by default
<anon>:2 #[allow(unstable)]
                 ^~~~~~~~
thread 'x509::X509Generator_0' panicked at 'test executable failed:

', ../src/librustdoc\test.rs:312


failures:
    bn::BigNum::abs_cmp_0
    bn::BigNum::checked_mod_add_0
    bn::BigNum::checked_shl1_0
    bn::BigNum::checked_shl1_1
    bn::BigNum::checked_sqr_0
    bn::BigNum::negate_0
    bn::BigNum::new_from_slice_0
    bn::BigNum::to_dec_str_0
    bn::BigNum::to_hex_str_0
    bn::BigNum::to_vec_0
    crypto::hash::Hasher_0
    crypto::hash::Hasher_1
    crypto::hmac::HMAC_0
    crypto::hmac::HMAC_1
    x509::X509Generator_0

test result: FAILED. 7 passed; 15 failed; 0 ignored; 0 measured

error: test failed
Segmentation fault
```

**Comment by sfackler** [comment]
Could you try running things in a debugger to get a backtrace of the location of the segmentation fault?

**Comment by CryptArchy** [comment]
Sure, give me a bit to figure this out.  I don't do systems programming regularly so this workflow is new to me.  Here's that initial run with gbd. Nothing to see yet, but I'll keep reading the docs and try to get better info.

```
$ gdb ~/Code/Rust/rust-openssl/openssl/target/debug/openssl-12ac8d499349ad2b.exe
...
Reading symbols from ~/Code/Rust/rust-openssl/openssl/target/debug/openssl-12ac8d499349ad2b.exe...done.
(gdb) run
Starting program: ~\Code\Rust\rust-openssl\openssl\target\debug\openssl-12ac8d499349ad2b.exe
[New Thread 2220.0x1410]

running 105 tests
[New Thread 2220.0x203c]
[New Thread 2220.0x1d40]
[New Thread 2220.0x1a8]
[New Thread 2220.0x23c0]
[New Thread 2220.0x1b3c]
[New Thread 2220.0x1dd4]
[New Thread 2220.0x1e80]
[New Thread 2220.0x2338]

Thread 5 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 2220.0x23c0]
0x0000000000000000 in ?? ()
```

**Comment by CryptArchy** [comment]
Trying to get a stack trace of the test output, and it doesn't go anywhere and suggests the stack is corrupt.  So I went back to my program and tried it, and was able to get a fairly detailed backtrace out of it.

```
(gdb) bt
#0  0x00000000007f3e74 in SSL_library_init ()
#1  0x0000000000741d73 in openssl_sys::init::{{closure}} ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-sys-0.7.13\src/lib.rs:398
#2  0x0000000000741d24 in std::sync::once::Once::call_once::_$u7b$$u7b$closure$u7d$$u7d$::h2634fec0f09afa93 () at ../src/libcore/cell.rs:881
#3  0x0000000000798b94 in std::sync::once::Once::call_inner::h92d827161dc52be3
    ()
#4  0x0000000000741bcc in std::sync::once::{{impl}}::call_once<closure> (
    self=0x994080 <openssl_sys::init::INIT::had6115b5c986203c>, f=...)
    at ../src/libstd\sync/once.rs:209
#5  0x0000000000741b36 in openssl_sys::init ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-sys-0.7.13\src/lib.rs:396
#6  0x000000000071585d in openssl::ssl::init::{{closure}} ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-0.7.13\src\ssl/mod.rs:58
#7  0x0000000000715834 in std::sync::once::Once::call_once::_$u7b$$u7b$closure$u7d$$u7d$::hc959285b4c1e7ef6 () at ../src/libcore/cell.rs:881
#8  0x0000000000798b94 in std::sync::once::Once::call_inner::h92d827161dc52be3
    ()
#9  0x00000000007156dc in std::sync::once::{{impl}}::call_once<closure> (
    self=0x994040 <openssl::ssl::init::INIT::h27efeed7b7420504>, f=...)
    at ../src/libstd\sync/once.rs:209
#10 0x0000000000715646 in openssl::ssl::init ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-0.7.13\src\ssl/mod.rs:57
#11 0x000000000071e5e9 in openssl::ssl::{{impl}}::new (method=Sslv23)
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-0.7.13\src\ssl/mod.rs:563
#12 0x000000000043c01b in hyper::net::openssl::{{impl}}::default ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\hyper-0.9.8\src/net.rs:657
#13 0x000000000043bfad in hyper::net::{{impl}}::default<hyper::net::openssl::OpensslClient,hyper::net::HttpConnector> ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\hyper-0.9.8\src/net.rs:573
#14 0x00000000004228e5 in hyper::client::pool::{{impl}}::new (config=...)
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\hyper-0.9.8\src\client/pool.rs:68
#15 0x00000000004073e1 in hyper::client::{{impl}}::with_pool_config (
    config=...)
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\hyper-0.9.8\src\client/mod.rs:123
#16 0x00000000004073a2 in hyper::client::{{impl}}::new ()
    at ~\.cargo\registry\src\github.com-1ecc6299db9ec823\hyper-0.9.8\src\client/mod.rs:118
#17 0x000000000040250f in sirocco::main () at src/main.rs:47
#18 0x00000000007aac8d in std::panicking::try::call::h852b0d5f2eec25e4 ()
#19 0x00000000007bb1c9 in __rust_maybe_catch_panic ()
#20 0x00000000007aaa85 in std::rt::lang_start::hfe4efe1fc39e4a30 ()
#21 0x000000000040659b in main ()
```

**Comment by sfackler** [comment]
It looks like this may be due to a mix of mingw and msvc libraries? http://mingw-users.1079350.n2.nabble.com/Crash-in-SSL-library-init-possible-linking-issue-td7578096.html

**Comment by CryptArchy** [comment]
Yes, I'm thinking that might be the issue so I'm trying to investigate that now.  I've been using mingw to build, trying to follow the advice of various readmes.  But I think that the windows build of OpenSSL for 64-bit doesn't contain the static mingw libraries.  (See the `VC` directory listed in my environment vars.)  Trying to see if I can build it in 32-bit.

**Comment by CryptArchy** [comment]
Alright, that fixed it! I'm not fully pleased by the outcome, as I'm now using a mix of gcc, mingw32, and the MSVC linker, but my program works and I got your openssl library to pass all it's tests.  Thank you very much @sfackler for taking the time to work with me on resolving my issue.  In case anyone comes across this ticket with similar problems, here's what ultimately worked for me:

1. Install MSYS2 and run by using `<msys-path>\msys2_shell.cmd -mingw32`
2. As per the readme instructions, run `pacman -S mingw-w64-i686-gcc` and `pacman -S mingw-w64-i686-toolchain`
3. As per the readme, install [Win32OpenSSL](http://slproweb.com/products/Win32OpenSSL.html). I used `Win32OpenSSL-1_0_2h.exe`
4. I used `rustup` to install this rust version `nightly-i686-pc-windows-gnu` (which currently resolves to `rustc 1.11.0-nightly (bb4a79b08 2016-06-15)`)
5. I used this script to set up environment variables

```
export OPENSSL_INCLUDE_DIR=/C/OpenSSL-Win32/include
export DEP_OPENSSL_INCLUDE=/C/OpenSSL-Win32/include
export OPENSSL_LIBS=ssleay32MT:libeay32MT
export OPENSSL_LIB_DIR=/C/OpenSSL-Win32/lib/VC
export OPENSSL_CONF=/C/OpenSSL-Win32/bin/openssl.cfg
export PATH=$PATH:~/.cargo/bin:/C/OpenSSL-Win32/bin
```

6. And then `cargo test` should work for this library. And `cargo run` works for my project that uses it. 

Thanks again for your help and your work on bringing openssl to Rust @sfackler ! I'll close this issue now, hopefully it helps someone else having these issues on Windows.
